File: trajectory_optimization/cpp/SatelliteTest.cpp
/*
 * Author: Nick Goodson
 * Jan 30th 2020
 *
*/

#include "iLQR.h"

using namespace Eigen;
using namespace std;


/**
 * Test the iLQR algorithm on a desired attitude change
 */
int main() {
    // Define sizes and sim params
    const int Nx = 7;  //(q, w)
    const int Nw = 3; // dimension of rotational dynamics
    const int Nu = 3;
    const double dt = 0.01;
    const double tol = 0.001;
    int N = 300;

    // Cost matrices
    MatrixXd Qwf = MatrixXd::Identity(Nw, Nw) * 100;  // Omega terminal cost
    MatrixXd Qw = MatrixXd::Identity(Nw, Nw) * 0.01;  // Omega cumulative cost
    MatrixXd R = MatrixXd::Identity(Nu, Nu) * 0.3;  // Control cumulative cost
    const double Qqf = 30;  // Attitude (quaternion) terminal cost

    // Initial and final states
    double theta0 = M_PI / 2;
    double thetaF = 3 * M_PI / 2;
    MatrixXd x0(Nx, 1);
    MatrixXd xg(Nx, 1);
    x0 << sin(theta0/2), 0, 0, cos(theta0/2), 0, 0, 0;
    xg << sin(thetaF/2), 0, 0, cos(thetaF/2), 0, 0, 0;

    // Outputs from iLQR (intitialized)
    MatrixXd xtraj = MatrixXd::Zero(Nx, N);
    xtraj(all, 0) = x0;
    MatrixXd utraj = MatrixXd::Zero(Nu, N-1);
    MatrixXd K = MatrixXd::Zero(Nu, Nx*(N-1));
    vector<double> Jhist;  // Size depends on how many iterations of while loop run. Use for testing only, don't implement on MCU

    // Call to iLQR function
    bool success = iLQRsimple(xg, Qw, R, Qwf, Qqf, dt, tol, xtraj, utraj, K, Jhist);
    cout << "Results: " << success << endl;

    // Write results to a csv file for comparison with MATLAB or python
    writeToFile(xtraj, utraj, Jhist);

    return 0;
}

File: trajectory_optimization/MATLAB/SatelliteTest.m
% Nick Goodson
% 5th Feb 2020
%
% Tests the quaternion iLQR function for controlling the
% attitude of a satellite

clear;
clc;
close all;

% Sim Params
N = 5000;  % num steps
dt = 0.01;
tol = 1e-3;
max_iters = 1000;
Nx = 7;
Nu = 3;

% Initial State
theta = pi/2;  % [rad]
q0 = [cos(theta/2); 0; 0; sin(theta/2)]  % 90 degree rotation about z-axis
w0 = [0; 0; 0];  % [rad/s]
x0 = [q0; w0];

% Goal state
theta_g = 3*pi/2;
qg = [cos(theta_g/2); 0; 0; sin(theta_g/2)]
wg = [0; 0; 0];
xg = [qg; wg];

% Initial Control trajectory
utraj0 = zeros(Nu, N-1);

% Cost matrices
% Cumulative
Q = zeros(Nx, Nx);
Qw = 0.1*eye(3);
Q(5:7, 5:7) = Qw;
R = 0.5*eye(3);

% Terminal
Qf= zeros(Nx, Nx);
Qwf = 1*eye(3);
Qf(5:7, 5:7) = Qwf;
Qqf = 200;  % Final cost for attitude error

[xtraj, utraj, K, Jhist, success] = iLQRsatellite(x0, xg, utraj0, Q, R, Qf, Qqf, dt, tol, max_iters);

% print final state
xtraj(:, end)

% Plot the quaternion time evolution
figure(1);
sgtitle("Quaternion")
subplot(4,1,1) 
plot(xtraj(1,:));
subplot(4,1,2)
plot(xtraj(2, :));
subplot(4,1,3)
plot(xtraj(3,:));
subplot(4,1,4)
plot(xtraj(4,:));

%plot the controls
figure(2)
subplot(3,1,1)
sgtitle("Controls")
plot(utraj(1,:));
subplot(3,1,2)
plot(utraj(2,:));
subplot(3,1,3)
plot(utraj(3,:));

%plot omega
figure(3)
sgtitle("Omega")
subplot(3,1,1)
plot(xtraj(5, :))
subplot(3,1,2)
plot(xtraj(6, :))
subplot(3,1,3)
plot(xtraj(7,:))

%plot the cost(
figure(4)
sgtitle("Cost")
plot(Jhist)






